pennylane/math/single_dispatch.py

```diff
@@ -0,0 +1,321 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Autoray registrations"""
+# pylint:disable=protected-access,import-outside-toplevel,wrong-import-position
+from importlib import import_module
+import numbers
+
+import autoray as ar
+import numpy as np
+
+
+def _i(name):
+    """Convenience function to import PennyLane
+    interfaces via a string pattern"""
+    if name == "tf":
+        return import_module("tensorflow")
+
+    if name == "qml":
+        return import_module("pennylane")
+
+    return import_module(name)
+
+
+# -------------------------------- NumPy --------------------------------- #
+from scipy.linalg import block_diag as _scipy_block_diag
+
+ar.register_function("numpy", "flatten", lambda x: x.flatten())
+ar.register_function("numpy", "coerce", lambda x: x)
+ar.register_function("numpy", "block_diag", lambda x: _scipy_block_diag(*x))
+ar.register_function("builtins", "block_diag", lambda x: _scipy_block_diag(*x))
+ar.register_function("numpy", "gather", lambda x, indices: x[np.array(indices)])
+
+
+def _scatter_element_add_numpy(tensor, index, value):
+    tensor[tuple(index)] += value
+    return tensor
+
+
+ar.register_function("numpy", "scatter_element_add", _scatter_element_add_numpy)
+
+
+# -------------------------------- Autograd --------------------------------- #
+
+
+ar.autoray._BACKEND_ALIASES["pennylane"] = "autograd"
+ar.autoray._MODULE_ALIASES["autograd"] = "pennylane.numpy"

```

Comment:
I don't think so, the key here is that the first is a backend alias, and the second is a module alias.

Reply:
I realised my last comment didn't really explain it very well:

- Autoray has the string 'autograd' hardcoded as a backend constant. When it inspects the module of our numpy tensors, however, it gets `"pennylane"`, so the backend alias 'redirects' to 'autograd'.
- The module alias instead means that when dispatching to autograd, it will instead call `qml.numpy.` rather than `autograd.numpy.`, to take into account our autograd modification.

Reply:
```suggestion
# When autoray inspects PennyLane NumPy tensors, they will be associated with
# the 'pennylane' module, and not autograd. Set an alias so it understands this is
# simply autograd.
ar.autoray._BACKEND_ALIASES["pennylane"] = "autograd"

# When dispatching to autograd, ensure that autoray will instead call
# qml.numpy rather than autograd.numpy, to take into account our autograd modification.
ar.autoray._MODULE_ALIASES["autograd"] = "pennylane.numpy"
```